xc7/utils/prjxray_db_cache.py

```diff
@@ -0,0 +1,66 @@
+"""
+This is a database cache. Its a kind of proxy object for Python sqlite3 connection
+which operates on a memory copy of the database.
+
+Upon object creation the database is "backed up" to memory. All subsequent
+operations are then pefromed on this copy which yields in performance increase.
+"""
+import sqlite3
+from progressbar.bar import ProgressBar
+
+# =============================================================================
+
+class DatabaseCache(object):
+
+    def __init__(self, file_name, read_only=False):
+
+        self.file_name = file_name
+        self.read_only = read_only
+        self.bar = None
+
+        # File URI
+        if self.read_only:
+            uri = "file:%s?mode=ro"  % file_name
+        else:
+            uri = "file:%s?mode=rwc" % file_name
+
+        # Open connections
+        self.memory_connection = sqlite3.connect(":memory:")
+        self.file_connection   = sqlite3.connect(uri, uri=True)
+
+        # Load the database
+        print("Loading database from '{}'".format(self.file_name))
+        self.file_connection.backup(self.memory_connection, pages=100, progress=self._progress)
+
+        self.bar.finish()
+        self.bar = None
+
+    def __del__(self):
+
+        # Write back only if not read-onlu
+        if not self.read_only:
+            print("Dumping database to '{}'".format(self.file_name))
+            self.memory_connection.backup(self.file_connection, pages=100, progress=self._progress)
+
+            self.bar.finish()
+            self.bar = None
+
+        # Close connections
+        self.memory_connection.close()
+        self.file_connection.close()
+
+    def _progress(self, status, remaining, total):
+        """
+        Prints database copy progress.
+        """
+        if self.bar is None:
+            self.bar = ProgressBar(max_value = total)
+        else:
+            self.bar.update(total - remaining)
```

Comment:
I believe this should just be update(remaining)?

Reply:
Not really, according to the example from progressbar documentation (https://progressbar-2.readthedocs.io/en/latest/) it is fine.
```
import time
import progressbar

with progressbar.ProgressBar(max_value=10) as bar:
    for i in range(10):
        time.sleep(0.1)
        bar.update(i)
```

It also works correctly.

Reply:
Yep, I'm used to a count up, remaining counts down.